### PR TITLE
Adding a thread safe RNG utility function

### PR DESCRIPTION
**Pull request by divyanshk**
This includes part of the changes in https://github.com/pytorch/pytorch/pull/161044/

When using PyTorch's DataLoader with thread-based workers, all worker threads share the same global random number generator (RNG) state. This creates a race condition: multiple threads may call random functions like torch.randint() or torch.rand() simultaneously, leading to non-reproducible results.

`torch.thread_safe_generator()` solves this by returning a thread-local generator when called from within a DataLoader thread worker.

This PR:
* we only include the utility public function to return the RNG. The RNG will be populated with the thread dataloader PR linked above. Right now this PR doesn't open any new functionality, the function will return `None` as RNG state isn't populated for thread workers (there are no thread workers right now - will land with PR#161044).
* landing this function separately to enable integration with Torchvision random transforms.
* Also, refactored `WorkerInfo` in `worker.py` to be a frozen dataclass.

### CODE REVIEW

File: torch/utils/data/_utils/worker.py [hunk 1]
```diff
@@ -73,29 +73,23 @@ def is_alive(self) -> bool:
...
-    def __setattr__(self, key, val) -> None:
-        if self.__initialized:
-            raise RuntimeError(
-                f"Cannot assign attributes to {self.__class__.__name__} objects"
-            )
-        return super().__setattr__(key, val)
-
-    def __repr__(self) -> str:
-        items = [f"{k}={getattr(self, k)}" for k in self.__keys]
-        return f"{self.__class__.__name__}({', '.join(items)})"
```
Comment (aelavender) [inline torch/utils/data/_utils/worker.py]: Weird.  Seems like all of this was just creating the same behaviour as `@dataclass(frozen=True)`.  Do you know if there's any particular reason why dataclass wasn't used here originally?  If not, it makes sense to use standard stuff.

File: torch/utils/data/_utils/worker.py [hunk 2]
```diff
@@ -73,29 +73,23 @@ def is_alive(self) -> bool:
...
+    rng: Optional["_RNG"] = None
```
Comment (aelavender) [inline torch/utils/data/_utils/worker.py]: Put `from __future__ import annotations` at the top of this file.  Then you don't need the quotes here.

File: torch/random.py [hunk 3]
```diff
@@ -215,3 +227,35 @@ def fork_rng(
...
+    """
+    # Lazy import to avoid circular dependency during torch module initialization
+    # torch.__init__ loads torch.random early, but torch.utils.data triggers
+    # torch.distributed which needs torch to be fully initialized
+    from torch.utils.data import get_worker_info
+
+    worker_info = get_worker_info()
+    if worker_info is not None and worker_info.rng is not None:
+        return worker_info.rng.torch_generator
+    return None
```
Comment (aelavender) [inline torch/random.py:270]: Hmmm... wondering if this should just return `default_generator` so that it always returns a generator?  (Similar idea as above re defaulting to the global RNGs).

File: torch/utils/data/_utils/worker.py (hunk 2 above)
Comment (aelavender) [inline torch/utils/data/_utils/worker.py]: I'm wondering if it would make sense to make this non-optional.  Could have it default construct with references to the global RNGs.  

Would potentially simplify some usage.  For example, instead of
`x = worker_info.rng.random_generator.random() if worker_info.rng else random.random()`
I could just write
`x = worker_info.rng.random_generator.random()`

WDYT?

File: test/test_dataloader.py [hunk 4]
```diff
@@ -1066,16 +1066,7 @@ def _test_worker_info_init_fn(worker_id):
...
     assert not hasattr(dataset, "value"), "worker_info should have correct dataset copy"
-    # test that WorkerInfo attributes are read-only
-    try:
-        worker_info.id = 3999
-    except RuntimeError as e:
-        assert str(e) == "Cannot assign attributes to WorkerInfo objects"
-    try:
-        worker_info.a = 3
-    except RuntimeError as e:
-        assert str(e) == "Cannot assign attributes to WorkerInfo objects"
```
Comment (aelavender) [inline test/test_dataloader.py]: Oh these are buggy, haha.  Completely don't do anything if the assignment doesn't throw.  

Fine to remove if we're using standard impl `@dataclass(frozen=True)`

File: torch/utils/data/_utils/worker.py (hunk 2 above)
Comment (divyanshk) [reply torch/utils/data/_utils/worker.py]: Good point let me think about it. The only reason i made it optional was to not change the behavior for multiprocessing path. Making it non-optional will definitely make it cleaner.

File: torch/random.py (hunk 3 above)
Comment (divyanshk) [reply torch/random.py:270]: I specifically want to return `None` here. The output of this will feed into the generator arg of random ops like `torch.randint(0, 10, (2,), generator=generator`. Yes, technically if we provide the default global generator as the argument it would behave the same way as passing in `None`, but idk if the former will lead to some weird edge case based on how users structure their code. RIght now all torchvision transforms take in None and I want to continue that for multiprocessing use-case.

If we would add the RNG for MP case, i will gate this on another WorkerInfo attribute like `worker_method` (which is something I'll be adding in the larger PR anyway).

File: torch/utils/data/_utils/worker.py (hunk 2 above)
Comment (divyanshk) [reply torch/utils/data/_utils/worker.py]: Okay so, I see that not populating this field for multiprocessing does feel incomplete. Right now for mp, we just [seed](https://github.com/pytorch/pytorch/blob/main/torch/utils/data/_utils/worker.py#L258-L265) the global RNG. We would need the actual generators to populate the multiprocessing case. Now, `torch.manual_seed` actually returns the generator so that is good, but for python's random I'd have to access the private object `random._inst`. Similarly for numpy it would be `np.random.mtrand._rand`. Hmm i think that should be ok. 

Also responded to similar concern above @aelavender.

File: torch/random.py [hunk 5]
```diff
@@ -215,3 +232,35 @@ def fork_rng(
         torch.set_rng_state(cpu_rng_state)
         for device, device_rng_state in zip(devices, device_rng_states):
             device_mod.set_rng_state(device_rng_state, device)
+
+
+def thread_safe_generator() -> Optional[torch.Generator]:
```
Comment (ramanishsingh) [inline torch/random.py:237]: Shall we add a test for this
`test_thread_safe_generator_reproducibility` and `test_thread_safe_generator_no_race_condition`.

File: torch/utils/data/_utils/worker.py [hunk 6]
```diff
@@ -70,32 +73,28 @@ def is_alive(self) -> bool:
             return not self.manager_dead
 
 
-_worker_info: Optional["WorkerInfo"] = None
+_worker_info: Optional[WorkerInfo] = None
 
 
+@dataclass(frozen=True)
```
Comment (ramanishsingh) [inline torch/utils/data/_utils/worker.py]: [AI told me this]

we can have `slots=True` here
We need Python 3.10 or higher.

https://www.attrs.org/en/stable/glossary.html#term-slotted-classes

example script to test speed up. Lower gain for just 8 num_workers, but still slightly more opt.

```
"""Simple benchmark: dataclass with vs without __slots__"""

import sys
import timeit
from dataclasses import dataclass


@dataclass(frozen=True)
class WithoutSlots:
    a: int
    b: int
    c: int


@dataclass(frozen=True, slots=True)
class WithSlots:
    a: int
    b: int
    c: int


# Memory comparison
obj1 = WithoutSlots(1, 2, 3)
obj2 = WithSlots(1, 2, 3)

size1 = sys.getsizeof(obj1) + sys.getsizeof(obj1.__dict__)
size2 = sys.getsizeof(obj2)

print("=== MEMORY ===")
print(f"Without slots: {size1} bytes")
print(f"With slots:    {size2} bytes")
print(f"Savings:       {size1 - size2} bytes ({100*(size1-size2)/size1:.0f}%)")

# Speed comparison
print("\n=== ATTRIBUTE ACCESS (1M iterations) ===")
t1 = timeit.timeit("obj.a; obj.b; obj.c", globals={"obj": obj1}, number=1_000_000)
t2 = timeit.timeit("obj.a; obj.b; obj.c", globals={"obj": obj2}, number=1_000_000)

print(f"Without slots: {t1:.3f}s")
print(f"With slots:    {t2:.3f}s")
print(f"Speedup:       {100*(t1-t2)/t1:.0f}%")

```

File: torch/utils/data/_utils/worker.py [hunk 7]
```diff
@@ -257,12 +272,24 @@ def _worker_loop(
...
+
         if HAS_NUMPY:
             np_seed = _generate_state(base_seed, worker_id)
             import numpy as np
 
             np.random.seed(np_seed)
+            numpy_generator = np.random.mtrand._rand
+
+        rng = _RNG(
+            random_generator=random_generator,
```
Comment (ramanishsingh) [inline torch/utils/data/_utils/worker.py]: dumb question
Are we using/accessing `random_generator` and `numpy_generator` anywhere?
Also , random_generator is set as random_generator = random._inst, which is a global generator, and might not be thread safe?

**Review by ramanishsingh — APPROVED** [review]
Overall LGTM, just 3 comments. :)

File: torch/utils/data/_utils/worker.py (hunk 6 above)
Comment (divyanshk) [reply torch/utils/data/_utils/worker.py]: Yeah i saw slots . let me add them in the next PR, thanks

File: torch/utils/data/_utils/worker.py (hunk 7 above)
Comment (divyanshk) [reply torch/utils/data/_utils/worker.py]: That is right. This part of the code is for multiprocess workers, so global generators is correct.

And, no we are not accessing random_generator and numpy_generator anywhere. I added this to ensure `rng` in `WorkerInfo` is non-optional and we have it populated for mp workers too.

File: torch/random.py (hunk 5 above)
Comment (divyanshk) [reply torch/random.py:237]: I have that in the threading worker PR.
